### PR TITLE
feat: update kagi mcp to MCP 2026-07-28

### PR DESCRIPTION
**Pull request by Microck**
## Summary

- Update `kagi mcp` to MCP `2026-07-28`
- Replace the legacy initialization handshake with `server/discover`
- Require per-request protocol metadata and client capabilities
- Add cache hints, tool annotations, structured content, and modern result types
- Return tool execution failures as `isError: true` results
- Add integration coverage and documentation

## Verification

Automated checks run by OpenAI Codex:

- [x] `cargo fmt --check`
- [x] `cargo clippy --all-targets --all-features --locked -- -D warnings`
- [x] `cargo test --workspace --all-targets --locked`
- [x] MCP integration tests: 30 passed

## Docs

- [x] Updated MCP command documentation
- [x] Updated `CHANGELOG.md`

## Auth / Secrets

- [x] No tokens, cookies, or local configuration secrets committed

## AI assistance disclosure

- `agent_name`: OpenAI Codex
- `agent_version`: codex-cli 0.147.0
- `model_used`: GPT-5
- `human_testing`: Maintainer confirmed the listed automated checks were personally run: `cargo fmt --check`, `cargo clippy --all-targets --all-features --locked -- -D warnings`, `cargo test --workspace --all-targets --locked`, and the MCP integration test subset
- `contribution_summary`: Updated the Kagi MCP server for the stateless MCP 2026-07-28 protocol.

## Summary by CodeRabbit

* **New Features**
  * Updated MCP support to protocol version `2026-07-28`.
  * Added server discovery, metadata validation, tool annotations, sorted tool listings, structured JSON results, and response caching metadata.
  * Improved tool visibility controls and clarified read-only and mutating tool behavior.

* **Bug Fixes**
  * Tool execution failures now return completed results marked as errors.
  * Malformed requests, unsupported versions, cursors, and arguments receive clearer protocol errors.
  * Removed the legacy initialization handshake.

* **Documentation**
  * Updated MCP guidance for the latest protocol, request requirements, discovery, and error handling.

<h3>Greptile Summary</h3>

The PR upgrades the stdio server to the stateless MCP `2026-07-28` protocol and modernizes tool discovery and results.
- Replaces initialization with `server/discover` and validates protocol metadata on each request.
- Adds cache hints, tool annotations, structured JSON content, and completed error results.
- Updates MCP documentation and integration coverage for the new protocol behavior.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge.

No blocking failure remains.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/main.rs | Implements stateless MCP discovery, metadata validation, annotated tool definitions, modern result handling, and format-aware structured content; both previously reported structured-content defects are resolved. |
| tests/integration-cli.rs | Adds protocol and structured-content integration coverage, including explicit JSON overriding a non-JSON server default. |
| docs/commands/mcp.mdx | Documents the new discovery flow, mandatory request metadata, cache behavior, tool exposure, and result semantics. |
| CHANGELOG.md | Records the MCP protocol upgrade, breaking handshake change, and revised tool-error behavior. |

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client
    participant Kagi as kagi mcp
    Client->>Kagi: server/discover + per-request metadata
    Kagi-->>Client: versions, capabilities, cache hints
    Client->>Kagi: tools/list + per-request metadata
    Kagi-->>Client: annotated tool catalog
    Client->>Kagi: tools/call + metadata and arguments
    alt Successful execution
        Kagi-->>Client: complete result + content + optional structuredContent
    else Tool execution failure
        Kagi-->>Client: complete result + isError: true
    else Invalid protocol request
        Kagi-->>Client: JSON-RPC error
    end
```

<sub>Reviews (4): Last reviewed commit: ["feat: update kagi mcp to MCP 2026-07-28"](https://github.com/microck/kagi-cli/commit/7669598b33f037cb8fac6185e2d2088243d8b839) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=53164186)</sub>

### CODE REVIEW

**Comment by Microck** [comment]
@coderabbitai review

**Comment by Microck** [comment]
@codex review

File: src/main.rs [hunk 1]
```diff
@@ -3156,56 +3186,237 @@ async fn run_mcp(args: McpArgs, profile: Option<&str>) -> Result<(), KagiError>
...
+        "resultType": "complete",
+        "content": [{ "type": "text", "text": text }],
+        "isError": is_error,
+    });
+    if !is_error {
+        let text = result["content"][0]["text"]
+            .as_str()
+            .expect("MCP text content should be a string");
+        if let Ok(structured_content) = serde_json::from_str::<Value>(text) {
+            result["structuredContent"] = structured_content;
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs:3413]: Fixed by gating structuredContent on the effective MCP output format and adding a regression test for JSON-looking pretty output.

File: src/main.rs [hunk 2]
```diff
@@ -3156,56 +3186,237 @@ async fn run_mcp(args: McpArgs, profile: Option<&str>) -> Result<(), KagiError>
...
+                },
+                _ => json_rpc_error(id, -32601, format!("Method not found: {method}")),
             },
-            _ => serde_json::json!({
-                "jsonrpc": "2.0",
-                "id": id,
-                "error": {"code": -32601, "message": format!("Method not found: {method}")},
-            }),
+            Err(error) => json_rpc_error_with_data(id, error.code, error.message, error.data),
         };
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs:3233]: Reviewed against the MCP 2026-07-28 schema. This finding is not valid for this protocol revision: DiscoverRequest.params uses RequestParams, whose _meta requires protocolVersion and clientCapabilities. The versioning rules require every request to carry its protocol version and require UnsupportedProtocolVersionError for an unsupported version. The existing regression test preserves that behavior, so no code change is needed.

File: src/main.rs [hunk 3]
```diff
@@ -3156,56 +3186,237 @@ async fn run_mcp(args: McpArgs, profile: Option<&str>) -> Result<(), KagiError>
...
+) -> Result<Value, McpProtocolError> {
+    let params = request
+        .get("params")
+        .and_then(Value::as_object)
+        .expect("validated MCP request params should be an object");
+    if params.get("cursor").is_some() {
+        return Err(McpProtocolError::invalid_params(
+            "MCP tools/list does not issue pagination cursors because the complete tool catalog fits in one response",
+        ));
+    }
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs]: Fixed by treating a null cursor as omitted and adding a regression test for cursor: null.

File: src/main.rs [hunk 4]
```diff
@@ -3468,19 +3679,124 @@ fn mcp_tool_definitions(config: &McpServerConfig) -> Value {
...
+            | "kagi_redirect_create"
+            | "kagi_redirect_update"
+            | "kagi_redirect_delete"
+            | "kagi_redirect_enable"
+            | "kagi_redirect_disable"
+            | "kagi_site_pref_set"
+            | "kagi_site_pref_remove"
+            | "kagi_cli"
+    )
+}
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs:3803]: Fixed by classifying kagi_assistant as mutating, hiding it by default, and routing it through the gated mutating path.

File: src/main.rs [hunk 5]
```diff
@@ -3468,19 +3679,124 @@ fn mcp_tool_definitions(config: &McpServerConfig) -> Value {
...
         "inputSchema": input_schema,
+        "annotations": mcp_tool_annotations(name),
     })
 }
 
+fn mcp_tool_annotations(name: &str) -> Value {
+    let read_only = !mcp_mutating_tool_name(name);
+    let mut annotations = serde_json::json!({
+        "readOnlyHint": read_only,
+        "openWorldHint": mcp_open_world_tool_name(name),
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs:3766]: Fixed by classifying both kagi_assistant and kagi_ask_page as mutating, hiding them by default, and routing them through the gated path.

File: src/main.rs [hunk 6]
```diff
@@ -3468,19 +3679,124 @@ fn mcp_tool_definitions(config: &McpServerConfig) -> Value {
...
+    });
+    if !read_only {
+        annotations["destructiveHint"] = serde_json::json!(mcp_destructive_tool_name(name));
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs]: Fixed by omitting destructiveHint when an operation is not known to be destructive, avoiding a false non-destructive assertion for overwrites.

File: src/main.rs [hunk 7]
```diff
@@ -3156,56 +3186,288 @@ async fn run_mcp(args: McpArgs, profile: Option<&str>) -> Result<(), KagiError>
...
+            let uses_template = arguments.get("template").and_then(Value::as_str).is_some()
+                && arguments.get("news").and_then(Value::as_bool) != Some(true);
+            Ok(is_json && !uses_template)
+        }
+        "kagi_batch_search" | "kagi_summarize" | "kagi_news" | "kagi_news_search"
+        | "kagi_ask_page" | "kagi_translate" | "kagi_fastgpt" | "kagi_enrich_web"
+        | "kagi_enrich_news" | "kagi_smallweb" | "kagi_cli" => Ok(is_structured_output_format(
+            &mcp_output_format(arguments, &default)?,
+        )),
+        _ => Ok(is_structured_output_format(&default)),
```
[the remark this answers is not in the document]
Comment (Microck) [reply src/main.rs:3473]: Fixed by treating the News metadata tools' per-call format as authoritative for structuredContent and adding regression coverage for explicit JSON output.
